Add option to change header ordering to place quoted headers first

scramsby opened the pull request.

scramsby commented:
All tests are now passing so this is ready for review.

kimgr commented on tests/cxx/quoted_headers_first.cc:
```diff
@@ -0,0 +1,36 @@
+//===--- prefix_header_operator_new.cc - test input file for iwyu ---------===//
```
`quoted_headers_first.cc`

scramsby replied:
Done.

kimgr commented on tests/cxx/quoted_headers_first.cc:
```diff
@@ -0,0 +1,36 @@
+#include "subfolder/indirect_subfolder.h"
+
+std::unique_ptr<IndirectSubfolderClass> CreateIndirectSubfolderClass()
+{
```
Opening brace goes on the preceding line

scramsby replied:
Done.

kimgr commented on tests/cxx/quoted_headers_first.cc:
```diff
@@ -0,0 +1,36 @@
+#include <list>
+#include <map>
+#include <memory>
+#include "subfolder/indirect_subfolder.h"
```
Hmm, I wonder if it would be beneficial to test this with an associated header, too? To see that associated headers (usually included with quotes) are sorted before other quoted headers. Also, that would provide an opportunity to make sure `--quoted_headers_first` works in header files too (but then you'd probably need to add another quoted include in this `.cc` file).

kimgr replied:
Also, might be a good idea to try this with a PCH.

scramsby replied:
Done, including both associated header and pch.

kimgr commented on iwyu_globals.cc:
```diff
@@ -94,6 +94,8 @@ static void PrintHelp(const char* extra_msg) {
          "   --no_comments: do not add 'why' comments.\n"
          "   --no_fwd_decls: do not use forward declarations.\n"
          "   --verbose=<level>: the higher the level, the more output.\n"
+         "   --quoted_headers_first: when sorting includes, place quoted\n"
```
"Quoted" is a quality of the `#include` line, not of the header.

All the other `LineSortOrdinal`s talk about "headers", but I think it makes sense to name this separately. The others use all sorts of heuristics to identify the kind of header, but this is really about the include line.

So I'd suggest `--quoted_includes_first`, and a thorough rename :-/

scramsby replied:
Done.

kimgr commented on iwyu_output.cc:
```diff
@@ -1778,6 +1778,7 @@ enum class LineSortOrdinal {
   PrecompiledHeader,
   AssociatedHeader,
   AssociatedInlineDefinitions,
+  QuotedHeader,
```
And this would then become `QuotedInclude`

scramsby replied:
Done.

kimgr commented on iwyu_output.cc:
```diff
@@ -1800,6 +1801,9 @@ LineSortOrdinal GetLineSortOrdinal(const OneIncludeOrForwardDeclareLine& line,
     return LineSortOrdinal::AssociatedHeader;
   }
+
+  if (GlobalFlags().quoted_headers_first && EndsWith(quotedInclude, "\""))
```
Oops, I just noticed now that `quotedInclude` doesn't follow the `hacker_case` naming convention. I'll push a commit to fix that, you may need to resolve conflicts when rebasing.

scramsby replied:
I can include in my fix if you prefer.

scramsby commented:
Will address your suggestions hopefully today. Agreed that a configuration mechanism would be a good idea at some point. Perhaps using the same serialization format as the .imp files would be a good way to go.

scramsby commented:
Changes made to address all previous iteration suggestions.

scramsby commented on iwyu.cc:
```diff
@@ -332,7 +332,7 @@ class BaseAstVisitor : public RecursiveASTVisitor<Derived> {
   bool TraverseStmt(Stmt* stmt) {
     if (current_ast_node_->StackContainsContent(stmt))
-      return true;               // avoid recursion
+      return true;               // avoid recursion 
```
Looks like a bad merge. Fixing.

kimgr commented on tests/cxx/quoted_includes_first.h:
```diff
@@ -0,0 +1,38 @@
+//===--- quoted_includes_first.cc - test input file for iwyu --------------===//
```
`quoted_includes_first.h`

scramsby replied:
Done.

kimgr commented on tests/cxx/quoted_includes_first.h:
```diff
@@ -0,0 +1,38 @@
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
```
I think we prefer old-school include guards, but for test headers we typically don't bother. Please remove the `#pragma` for now -- there are portability subtleties around `#pragma once` that I would prefer to keep out of the test suite.

scramsby replied:
Done.

kimgr commented on tests/cxx/quoted_includes_first.h:
```diff
@@ -0,0 +1,38 @@
+
+class quoted_includes_class {
+public:
+    static std::unique_ptr<IndirectSubfolderClass> CreateIndirectSubfolderClass();
```
Two-space indentation.

scramsby replied:
Done.

kimgr commented on tests/cxx/quoted_includes_first.h:
```diff
@@ -0,0 +1,38 @@
+#include <memory>
+#include "subfolder/indirect_subfolder.h"
+
+class quoted_includes_class {
```
Use PascalCase for class names. That said, I'm not sure the class adds anything here, the method might as well be a free function (?)

scramsby replied:
Removed class.

kimgr commented on tests/cxx/quoted_includes_first.cc:
```diff
@@ -0,0 +1,39 @@
+#include "quoted_includes_first.h"
+
+std::unique_ptr<IndirectSubfolderClass> quoted_includes_class::CreateIndirectSubfolderClass() {
+    return std::make_unique<IndirectSubfolderClass>();
```
Only two spaces indentation

kimgr replied:
Oh, hmm... `make_unique` is C++14. I think this is now Clang's default, but it could be that there are platforms where this is not true (the default has been different for different targets before). Could you dumb this down to `return std::unique_ptr<IndirectSubfolderClass>(new IndirectSubfolderClass);`? Thanks!

scramsby replied:
Done and done.

scramsby commented:
Will address today. I think I'll keep the quoted include rename as part of the single commit because it is new to this PR so not much value to show addressing PR feedback separately.

scramsby commented:
Done!

kimgr commented:
Beautiful, thank you!
